doc/decisions/DR016-One-to-many-result-paging.md

```diff
@@ -0,0 +1,669 @@
+# DR016 One-to-many result paging
+
+- **Status:** Proposed
+- **Impact:** Medium
+- **Driver:** @feltech
+- **Approver:** @elliotcmorris @foundrytom
+- **Contributors:** @feltech @elliotcmorris
+- **Informed:** @mattdaw
+- **Outcome:** Split batching from paging and page using a cursor object
+
+## Background
+
+The OpenAssetIO API contains functions that can return an arbitrarily
+long list of results.  In particular, the current core API has functions
+for retrieving a list of versions of a given entity, and retrieving a
+list of entities that are "related to" another entity in some way.
+
+Taking the related entities query as an example: this could be sibling
+or parent-child relationships (e.g. shot and sequence), or more loose
+relationships (e.g. the project file that was executed to generate a
+render).
+
+The current API design provides the host with a list of entity
+references that correspond to an input entity-relationship pair.
+
+There may be multiple entity-relationship pairs provided as input to
+a single API call, allowing the manager plugin to leverage batching
+optimizations. Note that, typically, batching primarily addresses
+network latency concerns, in addition to database query optimizations.
+
+In the most general case, the host cannot know the size of the list of
+related entity references that will be returned to it. This clearly
+poses a scalability issue.
+
+Database and web APIs have had to deal with this problem for decades,
+and invariably provide some mechanism of returning slices of result sets
+or "pages" that can be iterated through, lazily fetching chunks of data
+in a controlled way.
+
+In an ideal situation, requesting subsequent pages has performance
+benefits over repeating the same query twice (or more) with different
+range arguments, and yet work is performed lazily as pages are requested
+so that no unnecessary processing of unused pages is performed.  Also,
+pages should be stable, in that results are not skipped nor appear more
+than once on successive pages.
+
+Ultimately, the implementation must be provided by the manager plugin
+authors. The mechanism chosen in this DR can only present an abstract
+interface that the plugin must satisfy. As a result, characteristics
+such as page stability and performance cannot be guaranteed, but we can
+at least ensure that the chosen mechanism does not preclude such
+optimisations.
```

Comment:
[Minor][Consider]
Super subjective thing, but this reads to me almost like we're saying "Yes, managers _should_ be stable but we can't force them to be", when I think in actuality we mean more like "Manager may want to be stable, and we don't want to get in the way of that, whilst we also don't want force it."

Maybe?

```suggestion
such as page stability and performance cannot be guaranteed, and nor
do we desire to place such restrictions upon the implementation. However
we can at least ensure that the chosen mechanism does not preclude such
optimisations.
```

Reply:
Yeah I like it, will commit your suggestion as-is.